userguide/storage.rst

```diff
@@ -713,53 +713,70 @@ existing datasets to close them--the remaining
 
-Deduplication is the process of ZFS transparently reusing a single
-copy of duplicated data to save space. Depending on the amount of
-duplicate data, deduplicaton can improve storage capacity, as less
-data is written and stored. However, deduplication is RAM intensive. A
-general rule of thumb is 5 GB of RAM per terabyte of deduplicated
-storage. **In most cases, compression provides storage gains
-comparable to deduplication with less impact on performance.**
-
-In %brand%, deduplication can be enabled during dataset creation. Be
-forewarned that **there is no way to undedup the data within a dataset
-once deduplication is enabled**, as disabling deduplication has
-**NO EFFECT** on existing data. The more data written to a
-deduplicated dataset, the more RAM it requires. When the system starts
-storing the DDTs (dedup tables) on disk because they no longer fit
-into RAM, performance craters. Further, importing an unclean pool can
-require between 3-5 GB of RAM per terabyte of deduped data, and if the
-system does not have the needed RAM, it will panic. The only solution
-is to add more RAM or recreate the pool.
-**Think carefully before enabling dedup!**
-This `article
-<http://constantin.glez.de/blog/2011/07/zfs-dedupe-or-not-dedupe>`_
-provides a good description of the value versus cost considerations
-for deduplication.
-
-**Unless a lot of RAM and a lot of duplicate data is available, do not
-change the default deduplication setting of "Off".**
-For performance reasons, consider using compression rather than
-turning this option on.
-
-If deduplication is changed to *On*, duplicate data blocks are removed
-synchronously. The result is that only unique data is stored and
-common components are shared among files. If deduplication is changed
-to *Verify*, ZFS will do a byte-to-byte comparison when two blocks
-have the same signature to make sure that the block contents are
-identical. Since hash collisions are extremely rare, *Verify* is
-usually not worth the performance hit.
-
-.. note:: After deduplication is enabled, the only way to disable it
-   is to use the :samp:`zfs set dedup=off {dataset_name}` command
-   from :ref:`Shell`. However, any data that has already been
-   deduplicated will not be un-deduplicated. Only newly stored data
-   after the property change will not be deduplicated. The only way to
-   remove existing deduplicated data is to copy all of the data off of
-   the dataset, set the property to off, then copy the data back in
-   again. Alternately, create a new dataset with
-   :guilabel:`ZFS Deduplication` left disabled, copy the data to the
-   new dataset, and destroy the original dataset.
+#ifdef freenas
+.. index:: Deduplication
+.. _Deduplication:
+
+Deduplication
+^^^^^^^^^^^^^
+
+Deduplication ("dedup") is a process where ZFS aims to store a single copy
+of any data that is repeated in different blocks or files, rather than storing 
+identical copies each time the data occurs. Deduplication 
```

Comment:
Need to rework this to describe how ZFS computes a cryptographic hash of block data.

Reply:
For similar reasons I'm not happy again: this is a guide that has to cater to all users. Why exactly does an ordinary user need a tutorial about the technical mechanics of how dedup is done "under the radar", or a tutorial on cryptographic hashing, beyond being told that ZFS identifies repeated data and dedup's it? They don't, for this kind of guide. The only place this is relevant is when choosing dedup algorithm where the guide already explains all that a user will need to know, to make a choice. (If not, fix it there not here).

Reply:
I think this can be simplified to explain that the checksum of a block is calculated when the block is saved, then dedup just compares those checksums.

Reply:
@wblock - if it helped inform a user's choices I'd be all for it, but I just don't see any relevance yet. Maybe I'm missing something? Meaning - what impact does this knowledge (or its lack), concerning whether dedup is done via hashes, byte compare, or magic pixie dust, have on a user or their NAS decisions **here**? *It just doesn't impact users*. If the speed, or resource use, impacts users then we can say that, but even then, a knowledge of the technique ZFS uses "under the hood" has no impact, so it seems irrelevant. Are we sure there's a tangible need for this, that it's not just something like "a user just ought to know this detail otherwise how will they appreciate it" or similar?

Reply:
@wblock - if it helped inform a user's choices I'd be all for it, but I just don't see any relevance yet. Maybe I'm missing something? Meaning - what impact does this knowledge (or its lack), concerning whether dedup is done via hashes, byte compare, or magic pixie dust, have on a user or their NAS decisions **here**? *It just doesn't impact users*. If the speed, or resource use, impacts users then we can say that, but even then, a knowledge of the technique ZFS uses "under the hood" has no impact, so it seems irrelevant. Are we sure there's a tangible need for this, that it's not just something like "a user just ought to know this detail otherwise how will they appreciate it" or similar?